### ISSUE TITLE
Flyout Issue: "ArgumentOutOfRangeException" upon "await Shell.Current.GoToAsync()"

### ISSUE THREAD

**Issue by RodBarnes** [dotnet/maui#13932]
### Description

Running the app on Android emulator (both API 31 and API 33) runs with no issues.  Stop the app, switch to dark mode, then take the same actions and it throws the exception:

>System.ArgumentOutOfRangeException
>  Message=Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')

Once the exception is thrown, it continues to occur even after switching back to light mode. In summary, it is like the first time running the app on the Android emulator (or other targets), it runs fine as long as it is in light mode.  But the first time the target is switched to dark mode it will throw the exception and will continue to throw it thereafter regardless of light or dark mode.  And, as noted below, this morning it came up and worked fine in light mode.  But, again, once I switched to dark mode I was back in the same situation.  It is like whatever is happening is cached in some way but eventually clears out -- like how it worked this morning the first time.  But a restart of my dev system (see below), did not restore the working condition.

**Important notes**
1. This behavior of working and then not working after going to dark mode occurs with **no change in app code**; no change in XAML or C#.
2. I was in the middle of pre-release testing and this code has been working in both dark mode and light mode (with no exception thrown) for weeks now.  Everything was looking great but I then noted some changes I wanted to make to the color themes `App.xaml`.  Testing in light mode continued fine.  But, testing in dark mode brought doom.
3. Reverting `App.xaml` didn't make a difference; once the exception was thrown, it kept being thrown even after reverting those changes or anything else I tried.

I determined the exception was fired by the call to `Shell.Current.GoToAsync()` in `GoToImages()` of `DetailVM.cs`.  I put a breakpoint on that line of code and, stepping through the code, it stopped on that line.  I then stepped forward.  It immediately throws the exception and never even makes it to the constructor of `ImageListVM.cs` or of `ImageListPage.cs`.  That's the most I could determine: that it happens upon this `await Shell.Current.GoToAsync()` call in this specific place -- and only in that place.

The other uses of `Shell.Current.GoToAsync()` succeed with no Exception.  All of these actions use this method and work without throwing the exception:
- Menu > tap any page from flyout.
- Menu > tap About, then tap [Tester].
- Menu > tap any page from flyout, then tap [Filter].

I put in multiple break points trying to find where it was actually failing but it never made it to the constructor of the VM or of the page code-behind.  It hits the `Shell.Current.GoToAsync()` then throws the exception.  Other details of what I've investigated are covered in the "workaround " section, below.

So, I gave up and decided to try running on a completely different system (my laptop).  I cloned the repo, did clean & build, ran a debug session on API 31 emulator in light mode.  As I hoped (expected?) the app worked fine.  But after stopping the app, switching to dark mode, and running a debug session, the exception once again was thrown.  Stopped the app, switched to light mode, then ran a debug session and the exception continued to be thrown.

However, after giving up yesterday, I came back this morning to my development system, and -- surprise! -- the app was working in light mode again with no exception thrown.  But, moving to dark mode, it again threw the exception and going back to light mode, it continues to do so.

**Interesting note:**  When I debug the Windows version, it does **not** throw this Exception but, **instead**, falls into this code in `App.g.i.cs` and stops on the `System.Diagnostics.Debugger.Break()` within the line `if DEBUG && !DISABLE_XAML_GENERATED_BREAK_ON_UNHANDLED_EXCEPTION`:

```
    partial class App : global::Microsoft.Maui.MauiWinUIApplication
    {
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.UI.Xaml.Markup.Compiler"," 1.0.0.0")]
        private bool _contentLoaded;
        /// <summary>
        /// InitializeComponent()
        /// </summary>
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.UI.Xaml.Markup.Compiler"," 1.0.0.0")]
        [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
        public void InitializeComponent()
        {
            if (_contentLoaded)
                return;

            _contentLoaded = true;

            global::System.Uri resourceLocator = new global::System.Uri("ms-appx:///Platforms/Windows/App.xaml");
            global::Microsoft.UI.Xaml.Application.LoadComponent(this, resourceLocator);

#if DEBUG && !DISABLE_XAML_GENERATED_BINDING_DEBUG_OUTPUT
            DebugSettings.BindingFailed += (sender, args) =>
            {
                global::System.Diagnostics.Debug.WriteLine(args.Message);
            };
#endif
#if DEBUG && !DISABLE_XAML_GENERATED_BREAK_ON_UNHANDLED_EXCEPTION
            UnhandledException += (sender, e) =>
            {
                if (global::System.Diagnostics.Debugger.IsAttached) global::System.Diagnostics.Debugger.Break();
            };
#endif
        }
    }
```

### Steps to Reproduce

1. Clone the repo, build for Android API 31 or 33
2. Run a debug session with the emulator in light mode.
5. Tap a tile; e.g., "Abraham Lincoln Birthplace".
6. Tap the button [Images] and the list of images will display.
7. Close (stop) the app.
8. Switch to dark mode.
9. Run a debug session.
10. Tap a tile; e.g., "Abraham Lincoln Birthplace".
11. Tap the button [Images] and the exception will be thrown.
12. Close (stop) the app.
13. Switch to light mode.
14. Run a debug session.
15. Tap a tile; e.g., "Abraham Lincoln Birthplace".
16. Tap the button [Images] and the exception will still be thrown.

It will continue being thrown from this point forward.

### Link to public reproduction project repository

https://github.com/RodBarnes/NationalParks

### Version with bug

7.0 (current)

### Last version that worked well

7.0 (current)

### Affected platforms

Android, Windows

### Affected platform versions

Android API 31, 33; Windows 10, 11

### Did you find any workaround?

None found but following are the details of what I tried.

I confirmed that the content of the `Images List<Image>` is correct and valid -- so no weirdness in the data.

I tried removing the `parameter` dictionary from this `Shell.Current.GoToAsync()` call in `DetailVM` and removing the `QueryParameter` from the top of `ImageListVM`.  The exception is still thrown.

I tried creating a `NewDetailVM.cs` page using the same code from the original `DetailVM.cs`.  Nope, still thew the exception.

I tried creating a new page (`TestPage`) that does not use a `QueryParameter` -- just a new page with the default elements displayed and changing `DetailVM.GoToImages()` to use that page instead of 'ImageListPage'.  Nope, still threw the exception at the exact same spot.  (NOTE: I'll try again tomorrow morning -- assuming the condition again clears -- and see if, once the exception isn't being thrown in light mode if it continues working in dark mode with this simple page.)

I've done clean and rebuilds.  I've done all the usual; i.e., closed Visual Studio, closed the Emulator, killed all VS-related processes, restarted my computer, etc.  Didn't make any difference.

I checked out the last Stable release -- where this was **definitely** working -- and it still throws the Exception.

I went to a completely different system (my laptop) and cloned the repo.  Cleaned, build; debug session in light mode and the exception was **not** thrown.  Switched to dark mode and the exception was thrown.  Switched to light mode and the exception continued to be thrown.

The next morning, did clean & build and ran with emulator in light mode.  It all worked find without any exception!  I then closed the app, switched the emulator to dark mode, ran a debug session -- and the exception appeared.  I then closed the app, switched the emulator to light mode, ran a debug session -- and the errror still appeared.

Removed the frame style from `ImageListPage.xaml` and did a test in light mode:  Exception still thrown

Switched to API 33 emulator, clean & build; debug session in light mode: **No** exception thrown

API 33 emulator; debug session in dark mode: Exception thrown
API 33 emulator; debug session in light mode: Exception continues to be thrown.

Removed the label style from `ImageListPage.xaml`.

API 31 emulator; debug session in light mode (Note: Emulator reset during the deploy):  Exception thrown.

Manually did a factory reset on the API 31 emulator, started the emulator from the Device Manager; debug session in light mode: Exception was thrown.

Restored the frame and label styles to `ImageListPage.xaml`.

Closed the emulator, closed VS, restarted my computer; clean & rebuild, factory reset API 31 emulator; debug session in light mode: Exception thrown.

 #### UPDATE
I created a branch and removed all `AppThemeBinding` from `App.xaml` and `AppShell.xaml` in favor of using an explicit `StaticResource` color.  If I can get it to work (like tomorrow morning) in light mode, I'll trying running dark mode with this branch to see if the exception is still thrown.

**Comment by RodBarnes** [comment]
#### UPDATE
I tried again today -- hoping I'd get another refresh and the app would work in light mode.  Nope, still got the exception.

I created an entirely new emulator -- nothing had ever run on this emulator.  I cleaned & built the App from the NoDark tag (`App.xaml` and `AppShell.xaml` have NO `AppThemeBinding` specifications in them; i.e., the app is set to **ignore** dark mode).  Exception was still thrown.

**Comment by RodBarnes** [comment]
#### UPDATE
I just completely uninstalled VS, manually removed any left-over folders I found, and rebooted my system.  I let it set for 10 minutes an then installed VS 17.5.2.  I loaded the solution, let it restore things, then ran a debug session.  The exception was still thrown.

**Comment by RodBarnes** [comment]
#### UPDATE

1. Cloned the CommunityToolkit.Maui from GitHub.
2. Copied my solution to new area, removed NuGet reference for CommunityToolkit.Maui.
3. Added the CommunityToolkit.Maui and CommunityToolkit.Maui.Core projects to the copied solution, and added project references.
4. Clean and Build was successful.
5. Emulator was in light mode; ran debug session and app runs fine, no exception from tapping [Images].
6. Stopped app, switched emulator to dark mode, ran debug session and tapping [Images] produced the exception.
7. Stopped app, switched back to light mode, ran debug session and tapping [Images] still produces the exception.

**Comment by RodBarnes** [comment]
@jsuarezruiz Can you provide some direction on how I can further investigate this?  It appears I need to step into the `dotnet/maui` code but I'm not sure what I need to complete this.  I know I'd need to remove the libraries and replace it with references to the source so I cloned the dotnet/maui repo but haven't been successful getting the repo to build.

**Comment by RodBarnes** [comment]
Ok, lacking any response on this -- and realizing that this is happening on Windows, also -- I decided to try debugging there.  Under Windows, stepping through from a break on the `Shell.Current.GoToAsync()` call, it IMMEDIATLEY falls to this code in the `App` class:

```
#if DEBUG && !DISABLE_XAML_GENERATED_BREAK_ON_UNHANDLED_EXCEPTION
            UnhandledException += (sender, e) =>
            {
                if (global::System.Diagnostics.Debugger.IsAttached) global::System.Diagnostics.Debugger.Break();
            };
#endif
```

That there is a condition named `DISABLE_XAML_GENERATED_BREAK_ON_UNHANDLED_EXCEPTION` pretty clearly identifies that the problem is coming from the compiled XAML.

**Comment by RodBarnes** [comment]
@jsuarezruiz If you or someone could provide some info on what pieces of the `dotnet/maui` repository I need to build and reference, that would help me debug into that code to find where the error is happening.  To date, I still have not been able to get the `dotnet/maui` repository to build successfully.  I've run out of ideas on what to try in my own code -- see all that I've tried in the above comments.

**Comment by RodBarnes** [comment]
#### UPDATE
While the issue is still present, further investigation has shown that is has nothing to do with dark mode -- though it still could be related to the changes in App.xaml.  Continueing to investigate.

The first time [Images] are tapped from the first displayed list (whether Parks, or any other), the exception is thrown.  Once thrown, going back, selecting any list page (even the same list page) from the menu, will fix the issue and subsequent taps on [Images] does not throw the exception.

**Comment by RodBarnes** [comment]
I'm closing this as I've resolved it.  I don't understand why what I was doing was causing this issue but the change described here fixed the issue.

From somewhere, I'd read that `Shell.CurrentItem` could be set in `<Shell />` of the `AppShell.xaml` and was doing so in this manner:
```
<Shell xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
       xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
       xmlns:p="clr-namespace:NationalParks.Pages"
       x:Class="NationalParks.AppShell"
       FlyoutBackgroundColor="{AppThemeBinding Light={StaticResource Background}, Dark={StaticResource BackgroundDark}}"
       CurrentItem="{x:Reference StartItem}">
```

```
<FlyoutItem Title="Parks" Icon="tree_green">
    <ShellContent ContentTemplate="{DataTemplate p:ParkListPage}" x:Name="StartItem" />
</FlyoutItem>

```
This was working -- and had been working for a long time -- except for the recent appearance of this issue where it began throwing the `ArgumentOutOfRangeException`.  This exception was thrown after starting the app -- and it displays the initial page -- and then tapping an item in the `<CollectionView />` -- doing `GoToAsync(ItemDetailPage)` -- then tapping the [Images] button -- doing `GoToAsync(ImageListPage)`.

Investigation showed that, while it had displayed the `ItemListPage`, it wasn't all there.  For example, besides this exception, if I tapped "Menu > Parks" while the "Parks" page was already displayed, it would reload the page and everything would work fine after that -- including this exception would not be thrown.  So, something about using `CurrentItem` to specify the start page wasn't doing everything that happened if I tapped that page from the flyout.

By removing that XAML code and just allowing the flyout to default to the first page, it works and eliminates the throwing of this exception as well as the behavior of reloading the page when tapping the page on the flyout.

However, funny thing now is: If I attempt to set `CurrentItem` in `<Shell />`, it is no longer a recognized property in the displayed list of properties.  So, maybe something changed?
